lambda_builders/builder.py

```diff
@@ -0,0 +1,129 @@
+    """
+    Helps you build AWS Lambda functions.
+    """
+
+    # ./actions
+    _DEFAULT_BUILD_ACTIONS_DIR = os.path.join(os.path.dirname(__file__), "actions")
+
+    def __init__(self,
+                 build_actions_dir=None,
+                 entrypoint_file_name="entrypoint.py"):
+
+        """
+        Initialize the builder.
+
+        :type build_actions_dir: str
+        :param build_actions_dir:
+            Optional, path to the folder containing all the build actions. If this is not provided, then we will use
+            the default path.
+
+        :type entrypoint_file_name: str
+        :param entrypoint_file_name:
+            Optional, name of the entrypoint file that the builder should search for within each build action folder
+        """
+
+        self.build_actions_dir = build_actions_dir or self._DEFAULT_BUILD_ACTIONS_DIR
+        self.entrypoint_file_name = entrypoint_file_name
+
+    def build(self,
+              source_dir,
+              artifacts_dir,
+              manifest_path,
+              action_name,
+              runtime=None,
+              optimizations=None,
+              options=None):
+        """
+        Build a function at the given source location using the given build action.
+
+        :type source_dir: str
+        :param source_dir:
+            Path to a folder containing the source code
+
+        :type artifacts_dir: str
+        :param artifacts_dir:
+            Path to a folder where the built artifacts should be placed
+
+        :type manifest_path: str
+        :param manifest_path:
+            Path to the dependency manifest
+
+        :type action_name: str
+        :param action_name:
+            Name of the Build Action to use to build this Lambda function.
+
+        :type runtime: str
+        :param runtime:
+            Optional, name of the AWS Lambda runtime that you are building for. This is sent to the builder for
+            informational purposes.
+
+        :type optimizations: dict
+        :param optimizations:
+            Optional dictionary of optimization flags to pass to the build action. **Not supported**.
+
+        :type options: dict
+        :param options:
+            Optional dictionary of options ot pass to build action. **Not supported**.
+
+        :raises lambda_builders.exceptions.UnsupportedBuiltActionError
+        """
+
+        action_klass = self._load_action(action_name)
+        action = action_klass(source_dir,
+                              artifacts_dir,
+                              manifest_path,
+                              runtime=runtime,
+                              optimizations=optimizations,
+                              options=options)
+
+        # TODO: Verify source & artifacts directory & manifests exists
+
+        if not action.is_supported():
+            raise UnsupportedManifestError(action_name, manifest_path)
+
+        action.resolve()
+        action.compile()
+        action.copy_source()
```

Comment:
Is this maybe delegating too many tasks to one class? The way I am visualizing this, I think each action here in the resolve/compile/copy_source list should be a separate class that is delegated too by the builder. 

The class hierarchy would look like:

* BaseBuilder
** PythonBuilder

* BaseAction
** PythonDependencyBuildAction
** CopySourceAction

More subclasses are added per language, and each specific builder is essentially this:
```python
class BaseBuilder(object):
    def __init__(self, *args):
        self.actions = []

    def build(self):
        for action in self.actions:
            action.execute()


class PythonBuilder(BaseBuilder):
    def __init__(self):
        self.actions = [
            PythonDependencyBuildAction(<subset of *args that makes sense here>),
            CopySourceAction(<subset of *args that make sense here>),
        ]


class PythonDependencyBuildAction(BaseAction): 
    def __init__(self, artifact_dir_path, requirements_path, runtime, dependency_builder=None):
        self._artifact_dir_path = artifact_dir_path
        self._requirements_path = requirements_path
        self._runtime = runtime
        if dependency_builder is None:
            dependency_builder = PythonPipDependencyBuilder()
        self._dependency_builder = dependency_builder

    def execute(self):
        self._dependency_builder.build_dependencies(  # <-- this would be whatever the interface is i'm proposing in my PR.
            self._artifact_dir_path, 
            self._requirements_path, 
            self._runtime,
        ) 

```    
I like breaking this all out and making it clearly separated by language/tool, all separately testable and composable. New languages would be really easy to add by just adding the the `Actions` they need, and making a subclass of `Builder` that arranges them in the correct order. Makes python's "Special" case of no compile readily apparent too.

I think without this we are coupling the idea of how the build works to our builder, which if we are supporting x languages and y package managers I don't think is going to be very sustainable/flexible/testable. Downside is we are subprocessing for each action that can't be performed in pure python. 

What do you think? I could sketch this out in more concrete code if it's not clear.

Reply:
I love this! Makes it way more composable and testable. For out of process builder, we can spawn process once and pass JSONRPC commands for each action. 

I really like how actions can be skipped or reordered. 

I will modify the implementation to use this pattern.